### ISSUE TITLE
BatchNromalization after layer without Nonlinearity adds Nonlinearity

### ISSUE THREAD

**Issue by IstrateRoxana** [Lasagne/Lasagne#884]
Hi all, is this behaviour below intended? I specifically said nonlinearity=None to the Dense layer, but after i added the batch_norm i got a nonlinearity...

>>> from lasagne.layers import InputLayer, DenseLayer, batch_norm
>>> from lasagne.nonlinearities import tanh
>>> l1 = InputLayer((64, 768))
>>> l2 = batch_norm(DenseLayer(l1, num_units=500, nonlinearity=None))
>>> from lasagne.layers import get_all_layers
>>> [l.__class__.__name__ for l in get_all_layers(l2)]
['InputLayer', 'DenseLayer', 'BatchNormLayer', 'NonlinearityLayer']

Update:

I checked and the nonlinearity is actually linear. But is there a way of removing this useless layer?

**Comment by f0k** [comment]
Specifying `nonlinearity=None` in the `DenseLayer` constructor is actually just a shortcut for `nonlinearity=lasagne.nonlinearities.linear`: https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/dense.py#L79-L80
This avoids a case distinction in `get_output_for()` (https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/dense.py#L124) or anywhere else the `nonlinearity` attribute is accessed. But this also means any layer will look like it has a nonlinearity that needs to be pulled out when applying batch normalization.

> But is there a way of removing this useless layer?

You can write your own `batch_norm` convenience function that removes the layer if the nonlinearity is the identity function:
```python
def my_batch_norm(*args, **kwargs):
    layer = batch_norm(*args, **kwargs)
    if layer.nonlinearity is lasagne.nonlinearities.linear:
        layer = layer.input_layer
    return layer
```
I wouldn't want to change the existing `batch_norm` function because it might be useful that it always produces the same number of layers. The useless layer doesn't cost us anything in training (it does not alter the Theano expression). Hope this helps!
